Add blog post: Inside the LLM Call — GenAI Observability with OpenTelemetry (#9708)
open-telemetry/opentelemetry.io: content/en/blog/2026/genai-observability/index.md

---
title: 'Inside the LLM Call: GenAI Observability with OpenTelemetry'
linkTitle: GenAI Observability with OpenTelemetry

> date: 2026-04-21

Reviewer: ```suggestion
date: 2026-05-14
```

token counts, and durations are included. Enabling content capture populates
span attributes with full prompt messages, system prompts, tool schemas, tool
arguments, and tool results.

## Exploring GenAI telemetry

Any OTLP-compatible backend can receive GenAI telemetry. For this walkthrough,
we use the [Aspire Dashboard](https://aspire.dev/dashboard/overview/) — a free,
open source telemetry viewer that ships as a Docker container. It accepts OTLP
data directly and provides a built-in trace viewer, metrics explorer, and
structured logs page — no cloud account required. It is well suited for local
development and debugging of GenAI workloads.

Run the following Docker command to get started:

```sh
docker run --rm -p 18888:18888 -p 4317:18889 -p 4318:18890 -d --name aspire-dashboard \
    -e ASPIRE_DASHBOARD_UNSECURED_ALLOW_ANONYMOUS=true \
    mcr.microsoft.com/dotnet/aspire-dashboard:latest
```

The dashboard collects telemetry sent to `http://localhost:4318`, and you can
view telemetry by visiting `http://localhost:18888`.

> ℹ️ The dashboard requires authentication by default.
> Use `-e ASPIRE_DASHBOARD_UNSECURED_ALLOW_ANONYMOUS=true` to allow anonymous access

> > during local development.

Reviewer: Since the docker command includes the anonymous flag, I think we can just add this to the paragraph rather than make it a callout.

```suggestion
The dashboard collects telemetry sent to `http://localhost:4318`, and you can
view telemetry by visiting `http://localhost:18888`. The dashboard also requires authentication by default. Use `-e ASPIRE_DASHBOARD_UNSECURED_ALLOW_ANONYMOUS=true` to allow anonymous access during local development.
```

telemetry, every prompt, response, and tool call is visible at a glance.

### Explore metrics

Navigate to the **Metrics** page and select the `copilot-chat` service. The
GenAI metrics are prefixed with `gen_ai`:

- **`gen_ai.client.operation.duration`** — histogram of LLM call latencies.
  Filter by `gen_ai.request.model` to compare models.
- **`gen_ai.client.token.usage`** — histogram of token consumption. Filter by
  `gen_ai.token.type` to separate `input` from `output` tokens.

These metrics let you estimate per-request cost, catch token-hungry prompts
before they hit production, detect latency regressions, and monitor usage
patterns across models and agents.

![Dashboard metrics page showing GenAI metrics](dashboard-metrics.png)

## Beyond this demo

The GenAI semantic conventions are already in use today and under active
development — your feedback on real-world usage directly shapes what gets
standardized next.

- Test GenAI instrumentation in your own applications and
  [report issues](https://github.com/open-telemetry/semantic-conventions/issues).
- Set up a telemetry visualization tool for local debugging:
  - [Aspire Dashboard](https://aspire.dev/dashboard/overview/)

>   - [Grafana Tempo](https://grafana.com/docs/tempo/latest/set-up-for-tracing/)
>   - [Jaeger](https://www.jaegertracing.io/)
> - Join the
>   [GenAI Semantic Conventions and Instrumentation SIG](https://github.com/open-telemetry/community/blob/5125996b5d159ff9aaa906f9a25226a821dc7bed/projects/gen-ai.md)
>   discussions.

Reviewer: I think we can leave out the middle bullet altogether.

```suggestion
- Test GenAI instrumentation in your own applications and
  [report issues](https://github.com/open-telemetry/semantic-conventions/issues).
- Join the
  [GenAI Semantic Conventions and Instrumentation SIG](https://github.com/open-telemetry/community/blob/5125996b5d159ff9aaa906f9a25226a821dc7bed/projects/gen-ai.md)
  discussions.
```

Author: I think it's useful to have tool links here are as a call to action. It saves people having to go back to what they're previously read to get links for more info.

But I don't feel strongly about it. Up to you.

Reviewer: We generally only link once on a page, so I'm going to leave it out. Thanks!

that visibility. They standardize how GenAI operations are recorded — the model
being called, input and output token counts, and when opted in, the full content
of prompts, completions, tool calls, and tool results.

This post walks through:

- Exporting GenAI telemetry from an LLM-powered app.
- Configuring the [Aspire Dashboard](https://aspire.dev/dashboard/overview/), an
  observability tool optimized for local development, to receive and display
  that telemetry.
- Exploring GenAI traces, metrics, and events in Aspire's GenAI visualizer.

## Exporting GenAI telemetry

For this walkthrough, we use VS Code Copilot to generate telemetry, since most
developers already have it installed. However, many coding assistants support
monitoring with OpenTelemetry:

- [VS Code Copilot](https://code.visualstudio.com/docs/copilot/guides/monitoring-agents)
  emits traces, metrics, and events for every agent interaction.
- [OpenAI Codex](https://developers.openai.com/codex/config-advanced#observability-and-telemetry)
  exports structured log events and OTel metrics for API requests, tool calls,
  and sessions.
- [Claude Code](https://code.claude.com/docs/en/monitoring-usage) exports

>   metrics and log events via OTel, with trace support in beta.

Reviewer 2: should we expand the list to mention GitHub Copliot CLI (would the same setting apply as in vscode btw?) and Gemini CLI https://geminicli.com/docs/cli/telemetry/ ?

Author: The list isn't ment to be exhaustive.

Beyond monitoring the tools you already use, you can add OpenTelemetry to your
own GenAI-powered app to get insight into how it interacts with LLMs.

### Configure telemetry export

Telemetry export requires a few settings. For
[VS Code Copilot](https://code.visualstudio.com/docs/copilot/guides/monitoring-agents),
open Settings and search for `copilot otel`:

| Setting                                   | Description                          | Value                                                 |
| ----------------------------------------- | ------------------------------------ | ----------------------------------------------------- |

> | `github.copilot.chat.otel.enabled`        | Enable OTel emission                 | `true`                                                |

Reviewer 2: TIL! 🥳

| `github.copilot.chat.otel.captureContent` | Capture full prompt/response content | `true`                                                |
| `github.copilot.chat.otel.otlpEndpoint`   | OTLP collector endpoint              | `"http://localhost:4318"` (default, no change needed) |

By default, no prompt content or tool arguments are captured with GenAI
telemetry, as these can contain sensitive data. Only metadata like model names,
token counts, and durations are included. Enabling content capture populates
span attributes with full prompt messages, system prompts, tool schemas, tool
arguments, and tool results.

## Exploring GenAI telemetry with the Aspire Dashboard

> The [Aspire Dashboard](https://aspire.dev/dashboard/overview/) is a free, open

Reviewer 2: I think we should call out that it's not the only tool that can be used - you can see the data in your backend (if you configured the endpoint), in Jaeger if you spinned it or use some other local tools out there. And that we're using Aspire as an example (an amazing one for sure).

UPDATE: I think it called out below, but I'd prefer it to be higher when we just introduce Aspire

source telemetry viewer that ships as a Docker container. It accepts OTLP data
directly and provides a built-in trace viewer, metrics explorer, and structured
logs page — no cloud account required. It is well suited for local development
and debugging of GenAI workloads.

Run the following Docker command to get started:

```sh
docker run --rm -it -p 18888:18888 -p 4317:18889 -p 4318:18890 -d --name aspire-dashboard \

>     mcr.microsoft.com/dotnet/aspire-dashboard:latest

Reviewer 2: your choice but I usually just run it with DOTNET_DASHBOARD_UNSECURED_ALLOW_ANONYMOUS to avoid the login dance

Author: Security folks will be upset if the primary example leaves the dashboard unauthenticated.

I'll add it as a note below.

[GenAI telemetry visualizer](https://aspire.dev/dashboard/explore/#genai-telemetry-visualization)
parses these attributes and renders a chat-style view of the conversation,
showing system prompts, user messages, assistant responses, and tool call
arguments and results. Launch the visualizer by selecting the sparkle icon next
to a GenAI span.

![Aspire Dashboard GenAI telemetry visualizer showing a chat-style view of prompts and responses](dashboard-genai-visualizer.png)

No more guessing about LLM usage or digging through raw JSON — every prompt,
response, and tool call is visible at a glance.

### Explore metrics

Navigate to the **Metrics** page and select the `copilot-chat` service. The
GenAI metrics are prefixed with `gen_ai`:

- **`gen_ai.client.operation.duration`** — histogram of LLM call latencies.
  Filter by `gen_ai.request.model` to compare models.
- **`gen_ai.client.token.usage`** — histogram of token consumption. Filter by
  `gen_ai.token.type` to separate `input` from `output` tokens.

These metrics let you estimate per-request cost, catch token-hungry prompts
before they hit production, detect latency regressions, and monitor usage
patterns across models and agents.

> 

Reviewer 3: Add a picture showing off the metrics view if the data looks interesting?

Author: GenAI metrics reported by copilot don't include correct explicit bounds. I've reported it as a bug. But the data doesn't look that good in graphs.

> ## Beyond this demo

Reviewer 4: Is there a plan to update the opentelemetry-demo project?

https://github.com/open-telemetry/opentelemetry-demo/blob/main/src/llm/README.md?plain=1

Author: Not by me 😄 

Follow up on that separately from this blog post.

The [Aspire Dashboard](https://aspire.dev/dashboard/overview/) is ideal for
local debugging, but the same OpenTelemetry data can be sent to any
OTLP-compatible backend. Because the telemetry follows the
[GenAI semantic conventions](/docs/specs/semconv/gen-ai/), it is interoperable
with instrumentation from other AI tools and frameworks. Whether you are
monitoring a Python application using the
[OpenTelemetry OpenAI instrumentation](https://github.com/open-telemetry/opentelemetry-python-contrib/tree/6733459e4ee6a1d7a46025a71da6887f83491223/instrumentation-genai/opentelemetry-instrumentation-openai-v2),
or a .NET application with

> [Microsoft.Extensions.AI](https://learn.microsoft.com/en-us/dotnet/ai/microsoft-extensions-ai),

Reviewer 2: nit

```suggestion
[Microsoft.Extensions.AI](https://learn.microsoft.com/dotnet/ai/microsoft-extensions-ai),
```

> [Microsoft.Extensions.AI](https://learn.microsoft.com/en-us/dotnet/ai/microsoft-extensions-ai),

Reviewer 2: I think there are a few other native instrumentations these days - MCP in .net and python, I'll bring it up in todays GenAI call and ask people to share native instrumentations they are aware of.

Author: The list isn't ment to be exhaustive.

the same attributes and span structures appear in your observability backend.

## Get involved

The GenAI semantic conventions are in active development — your feedback on
real-world usage directly shapes what gets standardized next.

- Test GenAI instrumentation in your own applications and
  [report issues](https://github.com/open-telemetry/semantic-conventions/issues).
- Try the [Aspire Dashboard](https://aspire.dev/dashboard/overview/) for local
  OpenTelemetry debugging and [contribute to it](https://github.com/microsoft/aspire).
- Join the

>   [SIG GenAI Observability](https://github.com/open-telemetry/community/blob/5125996b5d159ff9aaa906f9a25226a821dc7bed/projects/gen-ai.md)

Reviewer 2: ```suggestion
  [GenAI Semantic Conventions and Insstrumentation SIG](https://github.com/open-telemetry/community/blob/5125996b5d159ff9aaa906f9a25226a821dc7bed/projects/gen-ai.md)
```

cSpell:ignore: genai
---

Your AI agent just took 45 seconds to answer a simple question. Was it the
model? A slow tool call? A retry loop? Every time an application calls an LLM, a
chain of model calls, tool invocations, and token exchanges happens behind the
scenes — and without observability, you are guessing.

The OpenTelemetry
[Semantic Conventions for Generative AI](/docs/specs/semconv/gen-ai/) give you
that visibility. They standardize how GenAI operations are recorded — the model
being called, input and output token counts, and when opted in, the full content
of prompts, completions, tool calls, and tool results.

This post walks through:

- Exporting GenAI telemetry from an LLM-powered app.
- Configuring the [Aspire Dashboard](https://aspire.dev/dashboard/overview/), an
  observability tool optimized for local development, to receive and display
  that telemetry.
- Exploring GenAI traces, metrics, and events in Aspire's GenAI visualizer.

## Exporting GenAI telemetry

For this walkthrough, we use VS Code Copilot to generate telemetry, since most
developers already have it installed. However, many coding assistants support
monitoring with OpenTelemetry:

> - [VS Code Copilot](https://code.visualstudio.com/docs/copilot/guides/monitoring-agents)

Reviewer 5: Same note as before, some of these may not be following OTel semconv. Might be worth adding a note to say as much. Hopefully they will start to converge in the future.

you can view telemetry by visiting `http://localhost:18888`.

The dashboard requires a login token on first access. Retrieve it from the
container logs.

![Docker container logs showing the Aspire Dashboard login URL with token](dashboard-login.png)

### Explore traces

GenAI operations from VS Code Copilot are now recorded and observable. Use AI in
the app and then open the **Traces** page in the dashboard. You will see entries
for each LLM interaction.

Selecting a trace reveals the span tree: the top-level `invoke_agent` span with
child `chat` spans for each LLM call and `execute_tool` spans for each tool
invocation.

![Aspire Dashboard showing the span tree for a GenAI trace](dashboard-spans.png)

The span details show GenAI semantic convention attributes:

- **`gen_ai.request.model`** — the model used (for example, `gpt-4o`).
- **`gen_ai.usage.input_tokens`** and **`gen_ai.usage.output_tokens`** — token
  counts for each LLM call.
- **`gen_ai.response.finish_reasons`** — why the model stopped generating (for
  example, `stop` or `tool_calls`).

When content capture is enabled, much more detail is added to the span, such as

> prompts and tool calls. These are serialized as JSON objects on the span, which

Reviewer 5: The GenAI semconv captures prompt/completion context via **logs** (`gen_ai.client.inference.operation.details`) and/or span attributes (`gen_ai.input.messages` / `gen_ai.output.messages`), not "JSON objects on the span."

I think this distinction matters more now that OTel is [deprecating span events](https://opentelemetry.io/blog/2026/deprecating-span-events/) in favour of logs (see [OTEP 4430](https://github.com/open-telemetry/opentelemetry-specification/pull/4430)).

If the tools described in the post differ from the semconv, it's worth calling that out explicitly so readers don't carry away the wrong mental model. I know the instrumentation libraries are a work in progress, so it might be worth noting what it should be vs what it might be now?

Author: I meant span attributes with JSON payloads. I'll improve the langauge.

No more guessing about LLM usage or digging through raw JSON — every prompt,
response, and tool call is visible at a glance.

### Explore metrics

Navigate to the **Metrics** page and select the `copilot-chat` service. The
GenAI metrics are grouped under the `gen_ai` meter:

- **`gen_ai.client.operation.duration`** — histogram of LLM call latencies.
  Filter by `gen_ai.request.model` to compare models.
- **`gen_ai.client.token.usage`** — histogram of token consumption. Filter by
  `gen_ai.token.type` to separate `input` from `output` tokens.

These metrics let you estimate per-request cost, catch token-hungry prompts
before they hit production, detect latency regressions, and monitor usage
patterns across models and agents.

## Beyond this demo

The [Aspire Dashboard](https://aspire.dev/dashboard/overview/) is ideal for
local debugging, but the same OpenTelemetry data can be sent to any
OTLP-compatible backend.

Because the telemetry follows the
[GenAI semantic conventions](/docs/specs/semconv/gen-ai/), it is interoperable
with instrumentation from other AI tools and frameworks. Whether you are
monitoring a Python application using the

> [OpenTelemetry OpenAI instrumentation](https://github.com/open-telemetry/opentelemetry-python-contrib/tree/6733459e4ee6a1d7a46025a71da6887f83491223/instrumentation-genai/opentelemetry-instrumentation-openai-v2),

Reviewer 5: Is there a specific reason why we're pointing to a commit instead of using main?

```suggestion
[OpenTelemetry OpenAI instrumentation](https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation-genai/opentelemetry-instrumentation-openai-v2),
```

Author: There is a markdown rule:

```
Run npm run check:markdown
  npm run check:markdown
  shell: /usr/bin/bash -e {0}

> check:markdown
> npx markdownlint-cli2 '*.md' '{content,projects}/**/*.md'

markdownlint-cli2 v0.22.0 (markdownlint v0.40.0)
Finding: *.md {content,projects}/**/*.md
Linting: 1482 file(s)
Summary: 2 error(s)
content/en/blog/2026/genai-observability/index.md:146 error gh-url-hash GitHub blob/tree URLs should use tags/releases or full 40-character commit hashes (and avoid default branch refs) [Use a tag/release ID or full 40-character commit hash instead of default branch ref "main" in GitHub URL] [Context: "https://github.com/open-telemetry/opente…/opentelemetry-instrumentation-openai-v2"]
content/en/blog/2026/genai-observability/index.md:159 error gh-url-hash GitHub blob/tree URLs should use tags/releases or full 40-character commit hashes (and avoid default branch refs) [Use a tag/release ID or full 40-character commit hash instead of default branch ref "main" in GitHub URL] [Context: "https://github.com/open-telemetry/community/blob/main/projects/gen-ai.md"]
Error: Process completed with exit code 1.
```

---
title: 'Inside the LLM Call: GenAI Observability with OpenTelemetry'
linkTitle: GenAI Observability with OpenTelemetry
date: 2026-04-21
author: '[James Newton-King](https://github.com/jamesnk) (Microsoft)'
issue: https://github.com/open-telemetry/opentelemetry.io/issues/9707
sig: SIG GenAI Observability

> cSpell:ignore: genai

Reviewer 4: Consider adding this at the repo level instead of file level

https://github.com/JamesNK/opentelemetry.io/blob/ee69d4623644fec2b4cc5e43ff0798f002cfe4ca/.cspell/all-words.txt
